core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java

```diff
@@ -1523,14 +1523,17 @@ public synchronized void setQueryCompleted()
             }
             queryCompleted = true;
             inputsQueue.clear();
+            outputTaskFailureListeners.clear();
             noMoreInputs = true;
         }
 
         public void addOutputTaskFailureListener(TaskFailureListener listener)
         {
             Map<TaskId, Throwable> failures;
             synchronized (this) {
-                outputTaskFailureListeners.add(listener);
+                if (!queryCompleted) {
```

Comment:
should we assert in `outputTaskFailed` that `!queryCompleted`?

Reply:
Where does the "ignoring" happen. Can you add a code comment why it is safe - for future readers?

Reply:
Methods like `io.trino.execution.scheduler.PipelinedStageExecution.PipelinedStageStateMachine#transitionToFailed` do handle out-of-bound notifications

Reply:
I've added 

```       
        public void outputTaskFailed(TaskId taskId, Throwable failure)
        {
            if (queryCompleted) {
                return;
            }
            List<TaskFailureListener> listeners;
```
to ignore task failures after query is completed.

Reply:
But now you just moved the problem one level higher. How do we know that 
```
               if (queryCompleted) {
                    // ignore late task failed events after query is completed
                    return;
                }
```

is right thing to do?

I am not saying it is not - but please explain in comment why we do not care about late events.

Reply:
Why not just feed the `failures` to the listener without adding the listener to outputTaskFailureListeners ?

Reply:
> Why not just feed the failures to the listener without adding the listener to outputTaskFailureListeners 

Yes, that's how it works after query is finished

Reply:
It is true for listeners already registered - not for listeners that are about to be registered.

Reply:
ok - never mind - read it incorrectly